app/projectsmodel.cpp

```diff
@@ -14,192 +8,582 @@
+
+void ProjectsModel::onAuthChanged()
+{
+  if ( !mBackend->userAuth() || !mBackend->userAuth()->hasAuthData() ) // user logged out, clear created and shared lists
+  {
+    if ( mModelType == CreatedProjectsModel || mModelType == SharedProjectsModel )
+    {
+      beginResetModel();
+      mProjects.clear();
+      endResetModel();
+    }
+  }
 }
 
-int ProjectModel::rowCount( const QModelIndex &parent ) const
+void ProjectsModel::setMerginApi( MerginApi *merginApi )
 {
-  Q_UNUSED( parent );
-  return mProjectFiles.count();
+  if ( !merginApi || mBackend == merginApi )
+    return;
+
+  mBackend = merginApi;
+  initializeProjectsModel();
 }
 
-QString ProjectModel::dataDir() const
+void ProjectsModel::setLocalProjectsManager( LocalProjectsManager *localProjectsManager )
 {
-  return mLocalProjects.dataDir();
+  if ( !localProjectsManager || mLocalProjectsManager == localProjectsManager )
+    return;
+
+  mLocalProjectsManager = localProjectsManager;
+  initializeProjectsModel();
 }
 
-QString ProjectModel::searchExpression() const
+void ProjectsModel::setModelType( ProjectsModel::ProjectModelTypes modelType )
 {
-  return mSearchExpression;
+  if ( mModelType == modelType )
+    return;
+
+  mModelType = modelType;
+  initializeProjectsModel();
 }
 
-void ProjectModel::setSearchExpression( const QString &searchExpression )
+QString ProjectsModel::modelTypeToFlag() const
 {
-  if ( searchExpression != mSearchExpression )
+  switch ( mModelType )
+  {
+    case CreatedProjectsModel:
+      return QStringLiteral( "created" );
+    case SharedProjectsModel:
+      return QStringLiteral( "shared" );
+    default:
+      return QStringLiteral( "" );
+  }
+}
+
+QStringList ProjectsModel::projectNames() const
+{
+  QStringList projectNames;
+  LocalProjectsList projects = mLocalProjectsManager->projects();
+
+  for ( const auto &proj : projects )
+  {
+    if ( !proj.projectName.isEmpty() && !proj.projectNamespace.isEmpty() )
+      projectNames << proj.id();
+  }
+
+  return projectNames;
+}
+
+void ProjectsModel::loadLocalProjects()
+{
+  if ( mModelType == LocalProjectsModel )
   {
-    mSearchExpression = searchExpression;
-    // Hack to model changed signal
     beginResetModel();
+    mergeProjects( MerginProjectsList(), Transactions() ); // Fills model with local projects
     endResetModel();
   }
 }
 
-bool ProjectModel::containsProject( const QString &projectNamespace, const QString &projectName )
+bool ProjectsModel::containsProject( QString projectId ) const
 {
-  return mLocalProjects.hasMerginProject( projectNamespace, projectName );
+  std::shared_ptr<Project> proj = projectFromId( projectId );
+  return proj != nullptr;
 }
 
-void ProjectModel::syncedProjectFinished( const QString &projectDir, const QString &projectFullName, bool successfully )
+std::shared_ptr<Project> ProjectsModel::projectFromId( QString projectId ) const
 {
-
-  // Do basic validity check
-  if ( successfully )
+  for ( int ix = 0; ix < mProjects.size(); ++ix )
```

Comment:
iterator for loop by more suitable probably since you do mProjects[ix] 6 times below